service/src/main/java/com/epam/esm/service/impl/UserDetailsServiceImpl.java

```diff
@@ -0,0 +1,50 @@
+package com.epam.esm.service.impl;
+
+import com.epam.esm.dao.UserDao;
+import com.epam.esm.entity.User;
+import com.epam.esm.exceptions.ExceptionMessageKey;
+import com.epam.esm.exceptions.ExceptionResult;
+import com.epam.esm.exceptions.IncorrectParameterException;
+import com.epam.esm.exceptions.NoSuchEntityException;
+import com.epam.esm.validator.UserValidator;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.stereotype.Service;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+
+@Service
+public class UserDetailsServiceImpl implements UserDetailsService {
+    private final UserDao userDao;
+
+    private static final String ROLE_PREFIX = "ROLE_";
+
+    @Autowired
+    public UserDetailsServiceImpl(UserDao userDao) {
+        this.userDao = userDao;
+    }
+
+    @Override
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
+        ExceptionResult exceptionResult = new ExceptionResult();
+        UserValidator.validateEmail(email, exceptionResult);
+        if (!exceptionResult.getExceptionMessages().isEmpty()) {
+            throw new IncorrectParameterException(exceptionResult);
+        }
+
+        Optional<User> user = userDao.findUserByEmail(email);
+        if (!user.isPresent()) {
+            throw new NoSuchEntityException(ExceptionMessageKey.USER_NOT_FOUND);
+        }
+
+        List<SimpleGrantedAuthority> authorities = Arrays.asList(new SimpleGrantedAuthority(ROLE_PREFIX
+                + user.get().getRole().toString()));
+
+        return new org.springframework.security.core.userdetails.User(user.get().getEmail(), user.get().getPassword(), authorities);
```

Comment:
Instead of creating new Spring Security User object here, would not be better if User entity class extends from org.springframework.security.core.userdetails.User it?